mappings/bnfdmd/data_downloader.py

```diff
@@ -0,0 +1,129 @@
+                continue
+            url = urljoin(self.url, a_tag["href"])
+            filename = Path(unquote(urlparse(url).path)).name
+            match = filename_re.match(filename)
+            if match:
+                inner_text = a_tag.text
+                while not (
+                    valid_from := datepat.search(inner_text.replace("\xa0", " "))
+                ):
+                    # sometimes "valid from" inner text is across adjacent elements
+                    inner_text += a_tag.next_sibling()[0].text
+                valid_from = datetime.strptime(valid_from.group(), datefmt)
+                matches.append((match.group("date"), url, filename, valid_from))
+
+        if not matches:
+            raise RuntimeError(f"Found no URLs matching {filename_re} at {self.url}")
+
+        # Sort by release date and get the latest
+        matches.sort()
+
+        return matches
+
+    def get_latest_release(self):
+        return self.get_releases()[-1]
+
+    def get_release_metadata(self, release):
+        datestamp, url, filename, valid_from = release
+
+        return {
+            "release": datestamp,
+            "valid_from": valid_from,
+            "url": url,
+            "filename": filename,
+            "release_name": datestamp,
+        }
+
+    def get_latest_release_metadata(self):
+        latest = self.get_latest_release()
+        return self.get_release_metadata(latest)
+
+    def download_release(self, release_name, valid_from):
+        """
+        Download a release that matches the specified release_name and valid_from values
+        """
+        logger.info(
+            "Attempting to download release from NHS BSA",
+            release=release_name,
+            valid_from=valid_from,
+        )
+
+        releases = self.get_releases()
+        release_metadata = (self.get_release_metadata(release) for release in releases)
+
+        match_found = False
+        for metadata in release_metadata:
+            if (
+                metadata.get("release_name") == release_name
+                and metadata.get("valid_from") == valid_from
+            ):
+                match_found = True
+                break
+
+        if not match_found:
+            raise ValueError(
+                f"No matching release found for release {release_name}, valid from {valid_from}"
+            )
+
+        local_download_filepath = Path(self.release_dir) / metadata["filename"]
+        self.get_file(metadata["url"], local_download_filepath)
+
+        return local_download_filepath
+
+    def get_file(self, url, filepath):
+        logger.info("Starting download", download_filepath=filepath)
+        with requests.get(url, stream=True) as resp:
+            resp.raise_for_status()
+            with open(filepath, "wb") as f:
+                for chunk in resp.iter_content(chunk_size=8192):
+                    f.write(chunk)
+        logger.info("Download complete", download_filepath=filepath)
```

Comment:
thought: It's not a big problem, but this code *might* be a candidate for factoring out both places it appears.

The code, neither here, nor in `TrudDownloader` never uses the `self` reference, so it could be a class static method, or a shared utility function.

That said, it was previously moved from being a function into the class: https://github.com/opensafely-core/opencodelists/commit/979d995aee8e377a47257323340f3245adae29b0, and it also brings up the question of "where does this live?".

**So having said all that, I'm not sure we need to do much about it, especially as some of the other code is duplicated too.**

But we could make it, and the other duplicate, static methods though, to make it simpler for the future reader.

Reply:
I'll raise a deck-scrubbing issue to refactor this and the TrudDownloader